packages/imt.sol/contracts/internal/InternalLazyIMT.sol

```diff
@@ -209,32 +215,48 @@ library InternalLazyIMT {
     ) internal view returns (uint256[] memory) {
         uint40 numberOfLeaves = self.numberOfLeaves;
         require(index < numberOfLeaves, "LazyIMT: leaf must exist");
-        require(depth > 0, "LazyIMT: depth must be > 0");
-        require(depth <= MAX_DEPTH, "LazyIMT: depth must be < MAX_DEPTH");
 
-        uint256[] memory proof = new uint256[](depth);
-        uint256 levelCount = self.numberOfLeaves;
+        uint8 targetDepth = 1;
+        while (uint40(2) ** uint40(targetDepth) < numberOfLeaves) {
+            targetDepth++;
+        }
+        require(depth >= targetDepth, "LazyIMT: proof depth");
+        // pass depth -1 because we don't need the root value
+        uint256[] memory _elements = new uint256[](depth);
+        _levels(self, numberOfLeaves, targetDepth - 1, _elements);
```

Comment:
not an expert in solidity. any benefit modifying the value like this instead of returning it?

Reply:
Yes, if we return an array from `_levels` then in this function we have to have 2 arrays in this function. We can't use the returned array as `_elements` directly because we can't control the size, and `_levels` might return an array that is smaller than we need, depending on the state of the tree.

It also removes a copy operation in the logic below in some cases. When I was returning the array from `_levels` the tests took longer on average than the original implementation, this change makes the tests faster. Test timing isn't a great metric for performance, but i _think_ the gas cost should also be lower because less memory is used with fewer copies.